src/renderer/components/ReactFlowBox.tsx

```diff
@@ -318,6 +335,225 @@ const ReactFlowBox = memo(({ wrapperRef, nodeTypes, edgeTypes }: ReactFlowBoxPro
         [createNode, wrapperRef.current, reactFlowInstance]
     );
 
+    const [connectingFrom, setConnectingFrom] = useState<OnConnectStartParams | null>();
+    const [connectingFromType, setConnectingFromType] = useState<string | null>();
+    const [isStoppedOnPane, setIsStoppedOnPane] = useState<boolean>(false);
+
+    useEffect(() => {
+        if (connectingFrom) {
+            const { nodeId, inOutId } = parseHandle(connectingFrom.handleId!);
+            const node = nodes.find((n) => n.id === nodeId);
+            if (node) {
+                const nodeSchema = schemata.get(node.data.schemaId);
+                if (connectingFrom.handleType === 'source') {
+                    const outputType = nodeSchema.outputs[inOutId]?.type;
+                    setConnectingFromType(outputType);
+                } else if (connectingFrom.handleType === 'target') {
+                    const inputType = nodeSchema.inputs[inOutId]?.type;
+                    setConnectingFromType(inputType);
+                } else {
+                    log.error(`Unknown handle type: ${connectingFrom.handleType!}`);
+                }
+            }
+        }
+    }, [connectingFrom]);
+
+    const [searchQuery, setSearchQuery] = useState<string>('');
+    const matchingNodes = useMemo(
+        () =>
+            getMatchingNodes(searchQuery, schemata.schemata).filter((node) => {
+                if (!connectingFrom || !connectingFromType) {
+                    return true;
+                }
+                if (connectingFrom.handleType === 'source') {
+                    return node.inputs.some((input) => {
+                        return connectingFromType === input.type && input.hasHandle;
+                    });
+                }
+                if (connectingFrom.handleType === 'target') {
+                    return node.outputs.some((output) => {
+                        return connectingFromType === output.type;
+                    });
+                }
+                log.error(`Unknown handle type: ${connectingFrom.handleType!}`);
+                return true;
+            }),
+        [connectingFrom, connectingFromType, searchQuery, schemata.schemata]
+    );
+    const byCategories = useMemo(() => getNodesByCategory(matchingNodes), [matchingNodes]);
+    const menuRef = useRef<HTMLDivElement>(null);
+    const menu = useContextMenu(() => (
```

Comment:
As I said, ReactFlowBox is already complex enough as is. Please move your new menu either into its own component or into its own hook.